### PR TITLE
Set up basics

### PR DESCRIPTION
**Pull request by dannyroberts**
I rebuilt master using `git subtree`, which has to be the right way give how much easier it was. This PR was then rebased onto that.

A result of this is that whereas I had originally envisioned having an `environments/` directory within this repo, `git subtree` really only lets you pull out a directory _as_ a repo, so now this repo _is_ the environments directory.

I actually think that's not so bad. It's a paradigm that allows people to start their own by first just creating there it and then at some point entering it and running `git init`, and then eventually pushing it somewhere. Not sure I'd actually recommend that pattern since you get a lot out of having something like a travis test run like we do in this repo, but it's an easy mental model. (Think of it like tracking your /etc/ folder with git or something like that.)

Any objections?

### CODE REVIEW

**Comment by millerdev** [comment]
Not sure I'm following exactly what's going on here. Are we expecting people to clone our environments repo? I don't think I'm familiar with `git subtree`, so maybe I need to learn a new workflow for it to make sense? I quickly scanned the beginning of [this blog post](https://www.atlassian.com/blog/git/alternatives-to-git-submodule-git-subtree), and noticed these drawbacks:

> - You must learn about a new merge strategy (i.e. git subtree).
> - Contributing code back upstream for the sub-projects is slightly more complicated.
> - The responsibility of not mixing super and sub-project code in commits lies with you.

I'm not sure what the first one means yet. The second sounds unfortunate if it means it will be more complicated to modify our environment files. The third may not matter, but could it possibly result in confusion for our partners if they don't understand what's going on? For example, if they accidentally make sloppy commits. (That's par for the course for people new to git or versioning systems in general.)

Can you outline the workflow showing what commands would be necessary to...

- make a change to an environment file (internal, dimagi)
- make a change to commcare-cloud (internal, dimagi)
- bootstrap and setup a new environments repo (external partner)
- make a change to an environment file (external partner)

Less important, but plausible, so maybe just a thought experiment:

- contribute bug fix or new feature to commcare-cloud (external partner)

**Comment by snopoke** [comment]
@millerdev what @dannyroberts was saying about using `git subtree` was related to how he populated this repo with our current environments from the `commcare-cloud` repo. He wasn't suggesting that a 3rd party would use it at all.

I'll give my understanding of how these would work and @dannyroberts can comment if he sees them differently:

> make a change to an environment file (internal, dimagi)

Assuming you're meaning a configuration change and not the result of a change in `commcare-cloud` which I address in the next one, the change would happen directly in this repo. In terms of updating a control server I think the workflow you'd go for is to just pull the updates from this repo since we'd have commcare-cloud installed via pip.

> make a change to commcare-cloud (internal, dimagi)

If the change required updating the environment files then you'd need to make a parallel PR in this repo. The way it's setup here we don't have a way to make breaking changes that would still allow the build to pass before the `commcare-cloud` changes had been merged. If we start versioning `commcare-cloud` then we'd be able to do that (but would then also need a way to do the reverse for the `commcare-cloud` build that tests against our env files: https://github.com/dimagi/commcare-cloud/pull/2000/files#diff-09e49f6dbb36856693060791c190e801R19)

> bootstrap and setup a new environments repo (external partner)

Could copy an existing set of env files or start from scratch. There's no need to clone / fork any repo or anything like that. You could copy the `.travis` file from here as well.

> make a change to an environment file (external partner)

Same as for us.

> contribute bug fix or new feature to commcare-cloud (external partner)

Same as for us except from a forked repo.

File: requirements.in [hunk 1]
```diff
@@ -0,0 +1,3 @@
+-e git+https://github.com/dimagi/commcare-cloud.git@dmr/prep-for-separate-env#egg=commcare-cloud
+nose
+parameterized
```
Comment (snopoke) [inline requirements.in:3]: we could make these extra requirements part of the commcare-cloud setup: https://setuptools.readthedocs.io/en/latest/setuptools.html#declaring-extras-optional-features-with-their-own-dependencies

File: requirements.in [hunk 2]
```diff
@@ -0,0 +1,3 @@
+-e git+https://github.com/dimagi/commcare-cloud.git@dmr/prep-for-separate-env#egg=commcare-cloud
```
Comment (snopoke) [inline requirements.in:1]: I assume we'd version this at some point right?

**Review by snopoke — APPROVED** [review]
Minor comments

File: requirements.in (hunk 1 above)
Comment (dannyroberts) [reply requirements.in:3]: Oh, hmm.  I think they're already under the `[test]` extra, so maybe I just need to add `[test]` to the end of the above line. I'll try it out

File: requirements.in (hunk 2 above)
Comment (dannyroberts) [reply requirements.in:1]: Yeah my idea of the progression is that once that PR is merged, we'll point this to master; and then once we're uploading versioned releases to pypi, we'll change this line to just `commcare-cloud` (or `commcare-cloud[test]`) and pin it in requirements.txt via `pip-compile`.

Could eventually use some revisiting as it's a little funny to have a requirements file here that's pretty parallel to the one in `commcare-cloud`, but that should be too much of an issue.

**Comment by dannyroberts** [comment]
Thanks for the explanation @snopoke, I agree with all that.

Sorry @millerdev for the confusion in my description. The final product will be that `commcare-cloud` no longer has an `environments/` directory at all; instead, the user will check out this repo, install `commcare-cloud`, and then point `COMMCARE_CLOUD_ENVIRONMENTS` to this repo, or run `manage-commcare-cloud configure` which will guide you through that. I think there'll be a fair amount to iron out (such as how this impacts the control init script, etc.), but it's all things that we fundamentally have to reconsider for anyone else to do this effectively—so in a way the same things that'll be painful about it are the reasons we're doing it.

**Comment by dannyroberts** [comment]
With simon's go ahead and because this repo isn't plugged into anything yet anyway, I'm going to merge

**Comment by dannyroberts** [comment]
PS. Since in the short interim, environments will be both here and in commcare-cloud's `environments/` dir, I will come up with a process for efficiently replicating any changes there to here, and I'll report it out so that others can follow it. (There will likely be a bit of a tail even after we remove environments from commcare-cloud since there may be open PRs.)

**Comment by millerdev** [comment]
Thanks for the clarification @dannyroberts. I still think it would be useful to outline in the form of _commands executed in the terminal_, what steps are necessary for each of the items at the end of my previous comment. I'm assuming the process for each of those will be different from the current process, and this is exactly the kind of change that might cause confusion and frustration from end users if it is not well documented how they are expected to change their usage of `commcare-cloud`.

**Comment by dannyroberts** [comment]
@millerdev yeah for sure. I definitely won't flip the switch on the team without first having those questions answered and written up in a canonical place.

**Pull request merged**